transports/quic/src/hole_punching.rs

```diff
@@ -0,0 +1,143 @@
+        let addr = self.addr;
+        let mut hole_punch_map = self.hole_punch_map.lock().unwrap();
+        if let Some(sender) = hole_punch_map.remove(&(addr, peer_id)) {
+            if let Err(connection) = sender.send((peer_id, connection)) {
+                Poll::Ready(Ok(connection))
+            } else {
+                Poll::Pending
+            }
+        } else {
+            Poll::Ready(Ok((peer_id, connection)))
+        }
+    }
+}
+
+impl FusedFuture for MaybeHolePunchedConnection {
+    fn is_terminated(&self) -> bool {
+        self.upgrade.is_terminated()
+    }
+}
+
+pub(crate) struct HolePuncher {
+    endpoint_channel: endpoint::Channel,
+    remote_addr: SocketAddr,
+    timeout: Delay,
+    interval: Delay,
+    message: Option<ToEndpoint>,
+}
+
+impl HolePuncher {
+    pub(crate) fn new(
+        endpoint_channel: endpoint::Channel,
+        remote_addr: SocketAddr,
+        timeout: Duration,
+    ) -> Self {
+        Self {
+            endpoint_channel,
+            remote_addr,
+            timeout: Delay::new(timeout),
+            interval: Delay::new(Duration::from_secs(0)),
+            message: None,
+        }
+    }
+}
+
+/// Never finishes successfully, only with an Err (timeout)
+impl Future for HolePuncher {
+    type Output = Error;
+
+    fn poll(mut self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Self::Output> {
+        loop {
+            match self.timeout.poll_unpin(cx) {
+                Poll::Ready(_) => return Poll::Ready(Error::HandshakeTimedOut),
+                Poll::Pending => {}
+            }
+
+            let message = match self.message.take() {
+                Some(m) => m,
+                None => {
+                    futures::ready!(self.interval.poll_unpin(cx));
+                    self.interval.reset(Duration::from_millis(
+                        rand::thread_rng().gen_range(10..=200),
+                    ));
+                    ToEndpoint::SendUdpPacket(quinn_proto::Transmit {
+                        destination: self.remote_addr,
+                        ecn: None,
+                        contents: rand::thread_rng()
+                            .sample_iter(distributions::Standard)
+                            .take(64)
+                            .collect(),
+                        segment_size: None,
+                        src_ip: None,
+                    })
+                }
+            };
+
+            match self.endpoint_channel.try_send(message, cx) {
+                Ok(Ok(())) => {
+                    // Message sent. Continue to register waker on `self.interval`.
+                    continue;
+                }
+                Ok(Err(m)) => {
+                    // Endpoint is busy. Try again later.
+                    self.message = Some(m);
+                }
+                Err(endpoint::Disconnected {}) => {
+                    return Poll::Ready(Error::EndpointDriverCrashed);
+                }
+            }
```

Comment:
I think this would be a lot easier to understand if it were an `async fn`. Isn't this essentially a loop that does:

```rust
loop {
	sleep(Duration::from_millis(rand::thread_rng().gen_range(10..=200))).await;
	self.endpoint_channel.send().await;
}
```

You'll have to create the async `send` function on `Channel` but it think this is worth it. We can then combine it with a `future::select` on the timeout and report the error accordingly.